### PR TITLE
[Fuchsia] Connect to the VM service using an ssh tunnel

### PR DESCRIPTION
**Pull request by zanderso**
Soon, I will land a change to the VM service to remove special cases for Fuchsia. This will cause the VM service on Fuchsia to act the same as everywhere else, in particular listening for connections only from localhost. This change prepares for that change by connecting to the VM service using an ssh tunnel rather than directly.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 1]
```diff
@@ -343,9 +363,16 @@ class FuchsiaReloadCommand extends FlutterCommand {
     return <String>[path, name];
   }
 
+  Future<List<_PortForwarder>> _forwardPorts(List<int> remotePorts) {
+    final String config = '$_fuchsiaRoot/out/$_buildType/ssh-keys/ssh_config';
```
Comment (rmacnak-google) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: As discussed offline, the configuration here seems to create the tunnel as a deamon that outlives the flutter tool command.

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 2]
```diff
@@ -82,61 +82,81 @@ class FuchsiaReloadCommand extends FlutterCommand {
...
+    try {
+      final List<int> servicePorts = forwardedPorts.map(
+          (_PortForwarder pf) => pf.port).toList();
+
+      if (_list) {
+        await _listVMs(servicePorts);
+        // Port forwarding stops when the command ends. Keep the program running
+        // until directed by the user so that Observatory URLs that we print
+        // continue to work.
+        printStatus('Press any key to exit.');
```
Comment (rmacnak-google) [inline packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: stdin seems to be left in line mode, so I actually have to hit Enter.

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 3]
```diff
@@ -82,61 +82,81 @@ class FuchsiaReloadCommand extends FlutterCommand {
 
   bool _list;
 
+  static final String ipv4Lookback = InternetAddress.LOOPBACK_IP_V4.address;
```
[the remark this answers is not in the document]
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 4]
```diff
@@ -148,7 +168,7 @@ class FuchsiaReloadCommand extends FlutterCommand {
     bool connected = true;
     Socket s;
     try {
-      s = await Socket.connect("$_address", port);
+      s = await Socket.connect(InternetAddress.LOOPBACK_IP_V4, port);
```
[the remark this answers is not in the document]
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart (hunk 1 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart [hunk 5]
```diff
@@ -372,22 +399,76 @@ class FuchsiaReloadCommand extends FlutterCommand {
...
+                                      int remotePort) async {
+    final int localPort = await _potentiallyAvailablePort();
+    if (localPort == 0) {
+      printStatus(
+          '_PortForwarder failed to find a local port for $address:$remotePort');
+      return new _PortForwarder._(0, null);
+    }
+    final List<String> command = <String>[
+        'ssh', '-F', sshConfig, '-nNT', '-L',
+        '$localPort:127.0.0.1:$remotePort', address];
```
[the remark this answers is not in the document]
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done

File: packages/flutter_tools/lib/src/commands/fuchsia_reload.dart (hunk 2 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/commands/fuchsia_reload.dart]: Done

**Review by rmacnak-google — APPROVED** [review]

**Pull request merged**